gensim/models/callbacks.py

```diff
@@ -0,0 +1,149 @@
+        self.coherence = coherence
+        self.texts = texts
+        self.window_size = window_size
+        self.topn = topn
+        self.logger = logger
+        self.viz_env = viz_env
+        self.title = title
+
+    def get_value(self, **kwargs):
+        # only one of the model or topic would be defined
+        self.model = None
+        self.topics = None
+        super(CoherenceMetric, self).get_value(**kwargs)
+        cm = gensim.models.CoherenceModel(self.model, self.topics, self.texts, self.corpus, self.dictionary, self.window_size, self.coherence, self.topn)
+        return cm.get_coherence()
+
+
+class PerplexityMetric(Metric):
+    def __init__(self, corpus=None, logger=None, viz_env=None, title=None):
+        self.corpus = corpus
+        self.logger = logger
+        self.viz_env = viz_env
+        self.title = title
+
+    def get_value(self, **kwargs):
+        super(PerplexityMetric, self).get_value(**kwargs)
+        corpus_words = sum(cnt for document in self.corpus for _, cnt in document)
+        perwordbound = self.model.bound(self.corpus) / corpus_words
+        return np.exp2(-perwordbound)
+
+
+class DiffMetric(Metric):
+    def __init__(self, distance="jaccard", num_words=100, n_ann_terms=10, normed=True, logger=None, viz_env=None, title=None):
+        self.distance = distance
+        self.num_words = num_words
+        self.n_ann_terms = n_ann_terms
+        self.normed = normed
+        self.logger = logger
+        self.viz_env = viz_env
+        self.title = title
+
+    def get_value(self, **kwargs):
+        super(DiffMetric, self).get_value(**kwargs)
+        diff_matrix, _ = self.model.diff(self.other_model, self.distance, self.num_words, self.n_ann_terms, self.normed)
+        return np.diagonal(diff_matrix)
+
+
+class ConvergenceMetric(Metric):
+    def __init__(self, distance="jaccard", num_words=100, n_ann_terms=10, normed=True, logger=None, viz_env=None, title=None):
+        self.distance = distance
+        self.num_words = num_words
+        self.n_ann_terms = n_ann_terms
+        self.normed = normed
+        self.logger = logger
+        self.viz_env = viz_env
+        self.title = title
+
+    def get_value(self, **kwargs):
+        super(ConvergenceMetric, self).get_value(**kwargs)
+        diff_matrix, _ = self.model.diff(self.other_model, self.distance, self.num_words, self.n_ann_terms, self.normed)
+        return np.sum(np.diagonal(diff_matrix))
+        
+
+class Callback(object):
+    def __init__(self, metrics):
+        # list of metrics to be plot
+        self.metrics = metrics
+
+    def set_model(self, model):
+        self.model = model
+        self.previous = None
+        # check for any metric which need model state from previous epoch
+        if any(isinstance(metric, (DiffMetric, ConvergenceMetric)) for metric in self.metrics):
+            self.previous = copy.deepcopy(model)
+            # store diff diagnols of previous epochs
+            self.diff_mat = Queue()
```

Comment:
As there could be any no. of diff/convergence metric input (for ex. with different distance measures), so by using queue, it won't need to keep track of their no. and epoch count, and simply depend on the sequence order.